src/app/rosetta/lib/block.ml

```diff
@@ -0,0 +1,261 @@
+open Core_kernel
+open Async
+open Models
+
+(* TODO: Populate postgres DB with at least one of each kind of transaction and then make sure ops make sense *)
+module User_command_info = struct
+  module Kind = struct
+    type t =
+      [`Payment | `Delegation | `Create_token | `Create_account | `Mint_tokens]
+  end
+
+  type t =
+    { kind: Kind.t
+    ; fee_payer: [`Pk of string]
+    ; source: [`Pk of string]
+    ; receiver: [`Pk of string]
+    ; fee_token: Unsigned.UInt64.t
+    ; token: Unsigned.UInt64.t
+    ; fee: Unsigned.UInt64.t
+    ; nonce: Unsigned.UInt32.t
+    ; amount: Unsigned.UInt64.t option
+    ; hash: string }
+
+  module Op = struct
+    type 'a t = {label: 'a; related_to: 'a option} [@@deriving eq]
+  end
+
+  let account_id (`Pk pk) token_id =
+    { Account_identifier.address= pk
+    ; sub_account= None
+    ; metadata= Some (Amount_of.Token_id.encode token_id) }
+
+  let to_operations block_creator_pk (t : t) : Operation.t list =
+    (* First build a plan. The plan specifies all operations ahead of time so
+     * we can later compute indices and relations when we're building the full
+     * models.
+     *
+     * For now, relations will be defined only on the two sides of a given
+     * transfer. ie. Source decreases, and receiver increases.
+    *)
+    let plan : 'a Op.t list =
+      (* there is always a fee transfer
+       * which has a increase in the receiver and decrease in the source (2 ops)
+       * *)
+      [ {Op.label= `Fee_payer_dec; related_to= None}
+      ; {Op.label= `Fee_creator_inc; related_to= Some `Fee_payer_dec} ]
+      @
+      (* When amount is not none, we move the amount from source to receiver *)
+      match t.amount with
+      | Some amount ->
+          [ {Op.label= `Payment_source_dec amount; related_to= None}
+          ; { Op.label= `Payment_receiver_inc amount
+            ; related_to= Some (`Payment_source_dec amount) } ]
+      | None ->
+          []
+      (* TODO: Add all other ops (conditionally, to the plan depending on the kind) *)
+    in
+    List.mapi plan ~f:(fun i op ->
+        let operation_identifier i =
+          {Operation_identifier.index= Int64.of_int_exn i; network_index= None}
+        in
+        let related_operations =
+          match op.related_to with
+          | Some relate ->
+              List.findi plan ~f:(fun _ a ->
+                  [%eq:
+                    [ `Fee_payer_dec
+                    | `Fee_creator_inc
```

Comment:
This name is a bit weird. I wonder if `Fee_receiver` may be better, especially since some of the fee-receivers are for proofs, not only for the block creator.

Reply:
Not for user commands though right? All fee receivers within user commands are the block creator (correct me if I'm wrong).

Reply:
No fees are received as part of user commands, so this should be dropped then.